### ISSUE TITLE
Implementation of Conductivity

### ISSUE THREAD

**Issue by ymahlau** [ymahlau/fdtdx#10]
To properly simulate metallic materials, conductivity needs to be implemented. This might make the inverse design with time-reversibility difficult, but it should at least be implemented as an optional feature for the forward simulation.

**Comment by txdai** [comment]
I would suppose that under the current time reversal implementation for gradient computation, using complex values for permittivities would already present some instabilities. Perhaps whenever metallic materials and dispersive materials exist, the solver can switch to the recursive halving algorithm for auto differentiation?

**Comment by ymahlau** [comment]
I believe Maxwell's equations are actually no longer time-reversible with metallic or dispersive materials, so 'difficult' may have been an understatement in my initial comment. You can always switch the [reversible_fdtd](https://github.com/ymahlau/fdtdx/blob/15b0769fc01a388fcf3409fffc78c1cf35185605/src/fdtdx/fdtd/fdtd.py#L16) with the [checkpointed_fdtd](https://github.com/ymahlau/fdtdx/blob/15b0769fc01a388fcf3409fffc78c1cf35185605/src/fdtdx/fdtd/fdtd.py#L321) function call to use the recursive halving algorithm. We will add an error message when trying to use reversible gradient computation whenever it is not possible to do so. 

Regarding complex permittivities, I did not test the stability yet. But, I think using jnp.complex64 data type instead of jnp.float32 should give similar stability. If you test that and get different results feel free to share this. It would be very interesting.

For the time plan, expect the implementation of conductivity to be done end of february

**Comment by simenhu** [comment]
Could you just employ the same checkpointing strategy used in the PML? Depending on what you simulate only a fraction of the materials used are actually conducting/elcectrical losses, so you could save the fields the same way you do with the PML, using the same mechanism, to make the reverse backsolve work and thereby get most of the memory benefits with he backsolve strategy

**Comment by ymahlau** [comment]
Yes, I am pretty sure this would work. I think there may also be a smarter way to go about this using a combination of the adjoint method and automatic differentation to implement a custom vjp. I will take a look into that. 

I just realized that the time plan in my last comment was way off, but the implementation should be done in the next few weeks.

**Comment by ymahlau** [comment]
After a bit of contemplation about Maxwell's equations, i managed to perform the inverse FDTD simulation by simply adding a corresponding gain term for conductive materials. This works extremely well and does not need any saving mechanism:

Forward Simulation:
https://github.com/user-attachments/assets/6775099a-05d2-49d5-9931-b69eec5a0ff0

Backward Simulation:
https://github.com/user-attachments/assets/4dbb4b56-f9ef-4aed-acfb-6e90ddf72724

**Comment by ymahlau** [comment]
Closed through #48
